src/core/symbology/qgssymbollayer.cpp

```diff
@@ -1000,20 +1005,55 @@ void QgsSymbolLayer::prepareMasks( const QgsSymbolRenderContext &context )
   }
 }
 
-void QgsSymbolLayer::installMasks( QgsRenderContext &context, bool recursive )
+bool QgsSymbolLayer::installMasks( QgsRenderContext &context, bool recursive, const QRectF &rect )
 {
+  bool res = false;
   if ( !mClipPath.isEmpty() )
   {
     context.painter()->save();
     context.painter()->setClipPath( mClipPath, Qt::IntersectClip );
+    res = true;
+  }
+  else if ( rect.isValid() )
+  {
+    // find just the clip geometries within the area the symbol layer will be drawn over
```

Comment:
It looks like pretty much the same piece of code of prepareMasks, can we factor it a bit in a method ?

Reply:
Ok, done